uwblueprint/humane-society
# F24/sayi/invite-user-endpoint

## Notion ticket link

[Invite User Endpoint + Email Link](https://www.notion.so/uwblueprintexecs/Task-Board-db95cd7b93f245f78ee85e3a8a6a316d)



## Implementation description
* Implemented generateSIgnInLink in authService
* Implemented sendInviteEmail in authService (uses generateSignInLink)
* Implemented /auth/invite-user endpoint in authRoutes.ts (sends invite email to newly created user with INVITED status)
* Installed firebase app in frontend and initialized it in firebase.js (using new frontend/.env secrets)
* Login process: checkIfSignInLink (Login.tsx) -> loginWithSignInLink (AuthAPIClient.ts) -> /auth/loginWithSignInLink (authRoutes.ts)
    * (Login.tsx) checkIfSignInLink calls authAPIClient.loginWithSignInLink if a user is not already logged in and firebase has confirmed that the url is a sign in link.
    * (AuthAPIClient.ts) loginWithSignInLink makes the call to the firebase app to signInWithEmailLink and then posts accessToken and refreshToken to /auth/loginWithSignInLink (note that this is how the regular login process works)
    * (AuthRoutes.ts) /auth/loginWithSignInLink sets refreshToken as cookie and returns data in the expected structure for login
    * (AuthAPIClient.ts) if the post to /auth/loginWithSignInLink is successful, stores data in localStorage as a json object and returns the data.
    * (Login.tsx) if the previous steps are successful, the authenticated_user is set successfully (login is completed)


## Steps to test
0. Might have to Install yarn in your frontend directory manually with "yarn add firebase" (https://classic.yarnpkg.com/en/package/firebase) if docker-compose up --build isn't working
1. Create user by posting to /user/ endpoint
<img width="934" alt="User created within firebase" src="https://github.com/user-attachments/assets/ab2f88df-dcf9-42c0-a45c-7e7c4f9f5bdd">

2. Test /auth/invite-user errors
    a. with an email not in firebase
<img width="1245" alt="Invalid email passed to :auth:invite-user" src="https://github.com/user-attachments/assets/484b96f3-ec35-4aad-abfb-c317184e68d0">
         b. with an email that doesn't have UserStatus.INVITED
<img width="1241" alt="non-invited user email is passed to :auth:invite-user" src="https://github.com/user-attachments/assets/0ad50228-48de-407a-b623-98d31627b738">

3. Send invite email to newly created user with /auth/invite-user (successful)
<img width="1245" alt="Valid email passed to :auth:invite-user" src="https://github.com/user-attachments/assets/a1dfab9d-0ab6-483b-a091-88c490138456">
<img width="602" alt="Email sent to user" src="https://github.com/user-attachments/assets/8dd44e9c-3935-4ef9-ae60-d6f333a52968">

4. Test login process by logging into an alternate account and then clicking sign-in link
(Login with sign-in process isn't executed and original user stays logged in. Note that the sign-in link isn't invalidated, it will work if you log out and then click it).
(Tested but no image, check that original user's authenticated_user remains in local storage). 

5. After logging out, click the sign-in link (successful)
<img width="1680" alt="Successful login with sign-in link" src="https://github.com/user-attachments/assets/80c25425-b144-470b-a639-2383d9bd1921">
<img width="938" alt="Email is verified after successful login" src="https://github.com/user-attachments/assets/f1240eeb-e71f-42d1-9607-d61bc563454e">

6. After successfully signing in with sign-in link, logout and click again (test error)
<img width="1680" alt="Sign-in link clicked again" src="https://github.com/user-attachments/assets/ca9f1324-784b-4c1a-95c6-f3fd708a24af">
<img width="1680" alt="Sign-in link fails, no user is set" src="https://github.com/user-attachments/assets/7b486745-02f6-4b4a-a264-7b68a7ce3cb1">

7. Send a new invite link to the user with INVITED status but wait for 6 hours before clicking it. Link should have expired (not tested, but it should have same output as 6).
 

## What should reviewers focus on?
* Is the login process implemented well (are there places where it passes too much/too little information, or are the functions too tightly coupled).
* Note: User is not updated to ACTIVE status yet (should it be part of this process)
* Side note for future: in step 2a, it returns a 500 error code when it should return a 404 NotFound, so the getUserByEmail function in userService isn't functioning as expected

## Checklist
